Fix synchronization in GRPCServer

vitlibar opened the pull request:
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

Changelog category:
- Bug Fix

Changelog entry:
Fix synchronization in GRPCServer
This PR fixes https://github.com/ClickHouse/ClickHouse/issues/27024

Commit:
Fix synchronization in GRPCServer

nikitamikhaylov commented:
What is really fixed in this PR? Also add a few comments like "We don't need to lock any mutex, because a predicate for a condition variable is atomic."

vitlibar commented:
> What is really fixed in this PR? Also add a few comments like "We don't need to lock any mutex, because a predicate for a condition variable is atomic."

I've changed this PR, now it provides a completely different solution, and also I've added a link to an issue in the description.

vitlibar commented on src/Server/GRPCServer.cpp:
```diff
@@ -480,6 +481,40 @@ namespace
+    private:
+        bool value;
+        mutable std::mutex mutex;
+        mutable std::condition_variable changed;
```
It appears that the previous solution (`std::atomic<bool> value` + null mutex) didn't work very well. Because `changed.wait(lock, predicate)` is the same as

```
while(!predicate())
  changed.wait();
```

and sometimes the following could happen:

```
Thread1 - running wait(): the predicate is checked, and it's false
Thread2 - running set(): value of the predicate is changed, now it's true, but Thread1 don't know yet
Thread2 - running set(): changed.notify_all() is called
Thread1 - running wait(): changed.wait() is called
```

Finally `Thread1` is waiting for `changed` to be notified, but it's never going to happen because it was notified before `changed.wait()` was called.

nikitamikhaylov replied:
Sure, this is default mistake everyone does using condition variables.
